### DISCUSSION TITLE
old versions of node.js

### DISCUSSION THREAD

**Discussion by essential-existence** [termux/termux-packages#7914] [Q&A] [answered]
Hello. Where can I get ready-made binaries or how to build past versions of node.js directly on the device. Especially v14, for "cdr / code-server".
Thank you in advance!

**Comment by thunder-coding** [comment]
You can override node version for code server using `FORCE_NODE_VERSION=<Your current major nodejs version> yarn global add code-server --ignore-engines`.

Old debs for nodejs can be found from Actions Tab on GitHub and downloading build arrtifacts. Although it is **not guaranteed to work always** as dependencies are updated on a regular basis, so you are likely to face unable to find shared library linker errors.

It is recommended to build old versions from source instead

**Reply by essential-existence** [reply]
> You can override node version for code server using `FORCE_NODE_VERSION=<Your current major nodejs version> yarn global add code-server --ignore-engines`.
> 
> Old debs for nodejs can be found from Actions Tab on GitHub and downloading build arrtifacts. Although it is **not guaranteed to work always** as dependencies are updated on a regular basis, so you are likely to face unable to find shared library linker errors.
> 
> It is recommended to build old versions from source instead

`FORCE_NODE_VERSION=17.0.1 yarn global add code-server --ignore-engines
yarn global v1.22.15
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
warning "code-server > @coder/logger@1.1.16" has unmet peer dependency "@google-cloud/logging@^4.5.2".
[4/4] Building fresh packages...
[1/3] ⢀ argon2
[-/3] ⢀ waiting...
error /data/data/com.termux/files/home/.config/yarn/global/node_modules/argon2: Command failed.
Exit code: 1
Command: node-pre-gyp install --fallback-to-build
Arguments:
Directory: /data/data/com.termux/files/home/.config/yarn/global/node_modules/argon2
Output:
node-pre-gyp info it worked if it ends with ok
node-pre-gyp info using node-pre-gyp@1.0.6
node-pre-gyp info using node@17.0.1 | android | arm64
node-pre-gyp info check checked for "/data/data/com.termux/files/home/.config/yarn/global/node_modules/argon2/lib/binding/napi-v3/argon2.node" (not found)
node-pre-gyp http GET https://github.com/ranisalt/node-argon2/releases/download/v0.28.2/argon2-v0.28.2-napi-v3-android-arm64-unknown.tar.gz
node-pre-gyp ERR! install response status 404 Not Found on https://github.com/ranisalt/node-argon2/releases/download/v0.28.2/argon2-v0.28.2-napi-v3-android-arm64-unknown.tar.gz
node-pre-gyp WARN Pre-built binaries not installable for argon2@0.28.2 and node@17.0.1 (node-v102 ABI, unknown) (falling back to source compile with node-gyp)
node-pre-gyp WARN Hit error response status 404 Not Found on https://github.com/ranisalt/node-argon2/releases/download/v0.28.2/argon2-v0.28.2-napi-v3-android-arm64-unknown.tar.gz
gyp info it worked if it ends with ok
gyp info using node-gyp@8.3.0
gyp info using node@17.0.1 | android | arm64
gyp info ok
gyp info it worked if it ends with ok
gyp info using node-gyp@8.3.0
gyp info using node@17.0.1 | android | arm64
gyp info find Python using Python version 3.10.0 found at "/data/data/com.termux/files/usr/bin/python3"
gyp http GET https://nodejs.org/download/release/v17.0.1/node-v17.0.1-headers.tar.gz
gyp http 200 https://nodejs.org/download/release/v17.0.1/node-v17.0.1-headers.tar.gz
gyp http GET https://nodejs.org/download/release/v17.0.1/SHASUMS256.txt
gyp http 200 https://nodejs.org/download/release/v17.0.1/SHASUMS256.txt
gyp info spawn /data/data/com.termux/files/usr/bin/python3
gyp info spawn args [
gyp info spawn args   '/data/data/com.termux/files/usr/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/data/data/com.termux/files/home/.config/yarn/global/node_modules/argon2/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/data/data/com.termux/files/usr/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/data/data/com.termux/files/home/.cache/node-gyp/17.0.1/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/data/data/com.termux/files/home/.cache/node-gyp/17.0.1',
gyp info spawn args   '-Dnode_gyp_dir=/data/data/com.termux/files/usr/lib/node_modules/npm/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=/data/data/com.termux/files/home/.cache/node-gyp/17.0.1/<(target_arch)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=/data/data/com.termux/files/home/.config/yarn/global/node_modules/argon2',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.'
gyp info spawn args ]
gyp info ok
gyp info it worked if it ends with ok
gyp info using node-gyp@8.3.0
gyp info using node@17.0.1 | android | arm64
gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
make: Entering directory '/data/data/com.termux/files/home/.config/yarn/global/node_modules/argon2/build'
  CC(target) Release/obj.target/libargon2/argon2/src/ref.o
  CC(target) Release/obj.target/libargon2/argon2/src/argon2.o
  CC(target) Release/obj.target/libargon2/argon2/src/core.o
  CC(target) Release/obj.target/libargon2/argon2/src/blake2/blake2b.o
  CC(target) Release/obj.target/libargon2/argon2/src/thread.o
  CC(target) Release/obj.target/libargon2/argon2/src/encoding.o
  AR(target) Release/obj.target/argon2.a
/data/data/com.termux/files/usr/bin/sh: 1: ar: not found
make: *** [libargon2.target.mk:147: Release/obj.target/argon2.a] Error 127
make: Leaving directory '/data/data/com.termux/files/home/.config/yarn/global/node_modules/argon2/build'
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/data/data/com.termux/files/usr/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:194:23)
gyp ERR! stack     at ChildProcess.emit (node:events:390:28)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:290:12)
gyp ERR! System Linux 4.14.141-18751193
gyp ERR! command "/data/data/com.termux/files/usr/bin/node" "/data/data/com.termux/files/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "build" "--fallback-to-build" "--module=/data/data/com.termux/files/home/.config/yarn/global/node_modules/argon2/lib/binding/napi-v3/argon2.node" "--module_name=argon2" "--module_path=/data/data/com.termux/files/home/.config/yarn/global/node_modules/argon2/lib/binding/napi-v3" "--napi_version=8" "--node_abi_napi=napi" "--napi_build_version=3" "--node_napi_label=napi-v3"
gyp ERR! cwd /data/data/com.termux/files/home/.config/yarn/global/node_modules/argon2
gyp ERR! node -v v17.0.1
gyp ERR! node-gyp -v v8.3.0
gyp ERR! not ok
node-pre-gyp ERR! build error
node-pre-gyp ERR! stack Error: Failed to execute '/data/data/com.termux/files/usr/bin/node /data/data/com.termux/files/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js build --fallback-to-build --module=/data/data/com.termux/files/home/.config/yarn/global/node_modules/argon2/lib/binding/napi-v3/argon2.node --module_name=argon2 --module_path=/data/data/com.termux/files/home/.config/yarn/global/node_modules/argon2/lib/binding/napi-v3 --napi_version=8 --node_abi_napi=napi --napi_build_version=3 --node_napi_label=napi-v3' (1)
node-pre-gyp ERR! stack     at ChildProcess.<anonymous> (/data/data/com.termux/files/home/.config/yarn/global/node_modules/@mapbox/node-pre-gyp/lib/util/compile.js:89:23)
node-pre-gyp ERR! stack     at ChildProcess.emit (node:events:390:28)
node-pre-gyp ERR! stack     at maybeClose (node:internal/child_process:1062:16)
node-pre-gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:301:5)
node-pre-gyp ERR! System Linux 4.14.141-18751193
node-pre-gyp ERR! command "/data/data/com.termux/files/usr/bin/node" "/data/data/com.termux/files/home/.config/yarn/global/node_modules/argon2/node_modules/.bin/node-pre-gyp" "install" "--fallback-to-build"
node-pre-gyp ERR! cwd /data/data/com.termux/files/home/.config/yarn/global/node_modules/argon2
node-pre-gyp ERR! node -v v17.0.1
node-pre-gyp ERR! node-pre-gyp -v v1.0.6
node-pre-gyp ERR! not ok
Failed to execute '/data/data/com.termux/files/usr/bin/node /data/data/com.termux/files/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js build --fallback-to-build --module=/data/data/com.termux/files/home/.config/yarn/global/node_modules/argon2/lib/binding/napi-v3/argon2.node --module_name=argon2 --module_path=/data/data/com.termux/files/home/.config/yarn/global/node_modules/argon2/lib/binding/napi-v3 --napi_version=8 --node_abi_napi=napi --napi_build_version=3 --node_napi_label=napi-v3' (1)`

**Reply by thunder-coding** [reply]
```
FORCE_NODE_VERSION=17 yarn global add code-server --ignore-engines
```

Only specify major node version. Not full

**Reply by essential-existence** [reply]
And yet the error is the same...

**Reply by ghost** [reply]
> AR(target) Release/obj.target/argon2.a /data/data/com.termux/files/usr/bin/sh: 1: ar: not found make: *** [libargon2.target.mk:147: Release/obj.target/argon2.a] Error 127 make: Leaving directory

Install `binutils`. The error tells you that utility `ar` is not found. Termux has been fully moved to LLVM/Clang and doesn't install `binutils` anymore as part of compiler toolchain.

**Comment by essential-existence** [comment]
Do you know resources of ready-made assemblies of previous versions of node for termux?

**Comment by essential-existence** [comment, marked as answer]
Thanks everyone for the help! Here's a partial solution based on your suggestions:

> pkg install binutils;
> v=$(node -v); v=${v#v}; v=${v%%.*};
> FORCE_NODE_VERSION="$v" yarn global add code-server --ignore-engines;
> code-server;

**Reply by BufuWinner** [reply]
Am I the only one having trouble with extensions? After installing code server, I can't add any of them (e.g. Python) both from the extension menu and manually. From what I can see, the errors are raised from Node, could it Be the version? In theory it works only on v14 and nothing else, so forcing it...
